Begin deprecating des3-cbc-sha1

greghudson opened the pull request:
During initial authentication, if des3-cbc-sha1 is used for either the
reply key or session key, display a warning that it will be
deprecated.

greghudson commented:
This approach mirrors the one used in the Apple fork of Heimdal.  That code base uses a more permanent code fixture, by adding a krb5_enctype_warning() function and a warning flag on enctypes.  I think we can just modify the code when we need to add or remove a warning.

This approach only finds users or initial ticket services (usually krbtgt/REALM) using triple-DES.  Services won't generally be detected.  We could add a check for the ticket encryption key, but I don't think it would be of much value.  We don't have a good way of displaying a warning for TGS requests (unless the request happens to be made with kvno, but that's not the common case).
a
My plan is to escalate to disabling with a setting to reenable (like allow_weak_crypto, but probably not the same variable) in either 1.20 or 1.21, and then remove support in some release after that.

frozencemetery reviewed (approved):
First, thanks for starting on this and laying out a timeline.  I was going to ask you about scope and what would be useful today, but that's all answered now.

I've checked that kinit and kadmin both trip the warning and a message is printed.  It might be nice to add a test for this, but I don't think I want to hold this up for that.

greghudson commented:
It appears that des3 is already deprecated according to the enctype table (and therefore in logs and kadmin output).  I changed the warning to say "will be desupported" (although I still plan to have some releases where it can be reenabled via a flag), and changed the commit summary.

frozencemetery commented on src/lib/krb5/krb/get_in_tkt.c:
```diff
@@ -1604,7 +1604,7 @@ warn_des3(krb5_context context, krb5_init_creds_context ctx,
 
     banner = _("Warning: encryption type des3-cbc-sha1 used for "
-               "authentication is weak and will be deprecated");
+               "authentication is weak and will be desupported");
```
I think this change is good in general but "desupported" is a bit unclear.  "will be nsupported in the future" is clearer, but long; maybe "will be removed"?

greghudson replied:
I'll change it to "will be disabled" since that is the next step.

greghudson commented:
I added a documentation commit.  I expect to remove admin/advanced/retiring-des.rst at some point, but that's out of scope.

frozencemetery commented on doc/admin/conf_files/kdc_conf.rst:
```diff
@@ -844,22 +844,21 @@ Encryption types
 camellia128-cts-cmac camellia128-cts                 Camellia-128 CTS mode with CMAC
-des3                                                 The triple DES family: des3-cbc-sha1
-aes                                                  The AES family: aes256-cts-hmac-sha1-96, aes128-cts-hmac-sha1-96, aes256-cts-hmac-sha384-192, and aes128-cts-hmac-sha256-128
+des3                                                 The triple DES family: des3-cbc-sha1aes                                                  The AES family: aes256-cts-hmac-sha1-96, aes128-cts-hmac-sha1-96, aes256-cts-hmac-sha384-192, and aes128-cts-hmac-sha256-128
```
This change seems accidental.

frozencemetery commented on doc/admin/enctypes.rst:
```diff
@@ -123,24 +123,83 @@ Enctype compatibility
+it.
+
+
+Migrating away from older encryption types
```
I like the step-by-step checklist feel of this.  Do you think it would make sense to link it from retiring-des.rst?

greghudson replied:
Since I plan to remove retiring-des.rst I don't think there's much value in adding the cross-reference.

frozencemetery replied:
Ah okay, wasn't aware that was a candidate for deletion.  If this is to be the replacement I am in favor :)

kaduk replied:
I am, of course, biased (being the main author of retiring-des.rst) but in some sense a lot of it is a very-expanded version of "changing the krbtgt key", and that part might still have value even once DES is well and truly gone.  That said, we have version control so nothing is permanently gone, and I can't ask anyone to sign up for doing the work of extracting out the generic bits.
